nemo/collections/llm/bert/model/base.py

```diff
@@ -0,0 +1,647 @@
+        forward_args["packed_seq_params"] = get_packed_seq_params(batch)
+
+    return model(**forward_args)
+
+
+def default_layer_spec(config: "BertConfig") -> ModuleSpec:
+    """
+    Return MCore layer spec based on the bert type.
+    For bert_type == 'megatron', use mcore's default layer spec;
+    For bert_type == 'huggingface', use Post-LayerNorm layer spec.
+    """
+    bert_type = config.bert_type
+    assert (
+        bert_type == 'megatron' or bert_type == 'huggingface'
+    ), f'Unknown bert type {bert_type}, supported type for bert model is: megatron, huggingface'
+    if HAVE_TE:
+        if bert_type == 'megatron':
+            return bert_layer_specs.bert_layer_with_transformer_engine_spec
+        else:
+            return bert_layer_with_transformer_engine_spec_postln
+
+    if bert_type == 'megatron':
+        return bert_layer_specs.bert_layer_local_spec
+    else:
+        return bert_layer_local_spec_postln
+
+
+@dataclass
+class BertConfig(TransformerConfig, io.IOMixin):
+    """ "Model config for BERT model. Adpated from megatron.core.models.bert.bert_model.BertModel"""
+
+    fp16_lm_cross_entropy: bool = False
+    parallel_output: bool = True
+    share_embeddings_and_output_weights: bool = True
+    position_embedding_type: Literal["learned_absolute", "rope"] = "learned_absolute"
+    rotary_base: int = 10000
+    rotary_percent: float = 1.0
+    seq_len_interpolation_factor: Optional[float] = None
+    seq_length: int = 512
+    attention_softmax_in_fp32: bool = False
+    masked_softmax_fusion: bool = True
+    deallocate_pipeline_outputs = True
+    make_vocab_size_divisible_by: int = 128
+
+    transformer_layer_spec: Union[ModuleSpec, Callable[["BertConfig"], ModuleSpec]] = default_layer_spec
+    forward_step_fn: Callable = bert_forward_step
+    data_step_fn: Callable = bert_data_step
+
+    bert_type: Literal["megatron", "huggingface"] = "megatron"
+    add_pooler: bool = True
+    bert_binary_head: bool = True
+    add_lm_head: bool = True
+
+    def configure_model(self, tokenizer) -> "MCoreBertModelWrapperWithPostLNSupport":
+        """Configure the BERT Model.
+        For bert_type == 'megatron', num_tokentypes in embedding is controlled by whether model has binary head.
+        For bert_type == 'huggingface', tokentypes embedding is always added with num_tokentypes = 2.
+        """
+        vp_size = self.virtual_pipeline_model_parallel_size
+        if vp_size:
+            p_size = self.pipeline_model_parallel_size
+            assert (
+                self.num_layers // p_size
+            ) % vp_size == 0, "Make sure the number of model chunks is the same across all pipeline stages."
+        from megatron.core import parallel_state
+
+        transformer_layer_spec = self.transformer_layer_spec
+        if not isinstance(transformer_layer_spec, ModuleSpec):
+            transformer_layer_spec = transformer_layer_spec(self)
+
+        breakpoint()
+        if self.bert_type == 'megatron':
```

Comment:
I recommend removing this if else, and directly get `num_tokentypes` from the config directly, or maybe only override it if `self.bert_binary_head` is set to False.  The reason is RoBERTA uses  `num_tokentypes=1`, and I assume we want to use the same base model for it